cilium/cilium
# RFC: Spiffe Integration

### What is this? 

This is a request for comments PR regarding the Cilium-SPIFFE integration that has been proposed in [this](https://docs.google.com/document/d/1QB3JIomGmEqg34sCQCmn9A-RqrCwBJMQgj7hoFkwTE0/edit#heading=h.lzba1zbo2xvl) design document. This PR covers the "Identity Generation Hardening" part of such proposal. We already have some prototypes of the "Upgrading connections to mTLS" part but we prefer start discussing only the former one to keep it more scoped.  

### How to test?

If you want to test it by yourself you can follow these steps. Otherwise you can watch [this](https://asciinema.org/a/421142) recorded demo.

0. Download, compile and install cilium-agent based on this PR
1. Relax labels validations by updating the CNP CRD. 
```
$ kubectl apply -f examples/crds/v2/ciliumnetworkpolicies.yaml
```
2. Add the `--enable-spiffe` flag to cilium-agent
3. Deploy Spire. I prepared a yaml manifest with all components needed to deploy it. It's based on the [spire k8s quickstart tutorial](https://github.com/spiffe/spire-tutorials/tree/master/k8s/quickstart) but uses a custom spire-agent image that implements a POC of the spire privileged API. 

```
$ kubectl apply -f https://gist.githubusercontent.com/mauriciovasquezbernal/820e8e11fa546c2b4e04a3a9105b47ed/raw/017ea2d66cdd1e7227ebde57f39d82ca9f1eb0f8/spire.yaml
```

4. Check that spire pods are running

```
$ kubectl get pods -n spire
NAME                READY   STATUS    RESTARTS   AGE
spire-agent-rqsp9   1/1     Running   0          14s
spire-server-0      1/1     Running   0          14s
```
5. Create some registration entries. [This ](https://gist.githubusercontent.com/mauriciovasquezbernal/820e8e11fa546c2b4e04a3a9105b47ed/raw/017ea2d66cdd1e7227ebde57f39d82ca9f1eb0f8/create_registration_entries.sh) creates some registration entries for testing. 

```
$ ./create_registration_entries.sh
```

6. Restart cilium-agent. (This is something that should be improved in the code to avoid this step)
```
$ sudo systemctl restart cilium
```

7.  Create some pods 

```
$ kubectl run podubuntu --image=ubuntu:latest -- sleep 1000000
$ kubectl run podpraqma --image=praqma/network-multitool:latest -- sleep 100000
```

8. Restrict all egress traffic 

``` 
$ kubectl apply -f https://gist.githubusercontent.com/mauriciovasquezbernal/820e8e11fa546c2b4e04a3a9105b47ed/raw/017ea2d66cdd1e7227ebde57f39d82ca9f1eb0f8/deny-all-egress.yaml
```

9. Check that ping is failing 

```
$ kubectl exec podpraqma -- ping <podubuntu IP>
```
10. Deploy image based policy that allows those two pods to talk to each other

```
$ kubectl apply -f https://gist.githubusercontent.com/mauriciovasquezbernal/820e8e11fa546c2b4e04a3a9105b47ed/raw/017ea2d66cdd1e7227ebde57f39d82ca9f1eb0f8/image-based.yaml
```
11. Ping now works

```
$ kubectl exec podpraqma -- ping <podubuntu IP>
PING 10.11.0.173 (10.11.0.173) 56(84) bytes of data.
64 bytes from 10.11.0.173: icmp_seq=1 ttl=63 time=0.124 ms
```

### TODO
- [ ] Update code once Spire "privileged" PR is ready (https://github.com/spiffe/spire/issues/2246 & https://github.com/spiffe/spire-api-sdk/pull/7)
- [ ] Tests (both unit and integration)
- [ ] How to deploy spire. Create helm templates?, leave it up to the user?
- [ ] Improve label validation (undo validation relax done for this POC)
- [ ] How to initialize the spiffe watcher in a lazy way (avoid trying to connect to it when spire is not deployed yet)
- [ ] Rebase once https://github.com/cilium/cilium/pull/15123 is merged
- [ ] Rebase once https://github.com/cilium/cilium/pull/16585 is merged 

ref https://github.com/cilium/cilium/issues/4016

cc @nyrahul @jrajahalme @jrfastab @navarrothiago @rscampos
